src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -666,27 +726,178 @@ public bool Remove(TKey key, out TValue value)
+                else
+                {
+                    if (typeof(TKey) == typeof(byte))
+                    {
+                        return (byte)(object)key0 == (byte)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(sbyte))
+                    {
+                        return (sbyte)(object)key0 == (sbyte)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(ushort))
+                    {
+                        return (ushort)(object)key0 == (ushort)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(short))
+                    {
+                        return (short)(object)key0 == (short)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(char))
+                    {
+                        return (char)(object)key0 == (char)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(uint))
+                    {
+                        return (uint)(object)key0 == (uint)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(int))
+                    {
+                        return (int)(object)key0 == (int)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(ulong))
+                    {
+                        return (ulong)(object)key0 == (ulong)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(long))
+                    {
+                        return (long)(object)key0 == (long)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(IntPtr))
+                    {
+                        return (IntPtr)(object)key0 == (IntPtr)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(UIntPtr))
+                    {
+                        return (UIntPtr)(object)key0 == (UIntPtr)(object)key1;
+                    }
+                    else if (typeof(TKey) == typeof(Guid))
+                    {
+                        return (Guid)(object)key0 == (Guid)(object)key1;
+                    }
+                    else
+                    {
+                        return EqualityComparer<TKey>.Default.Equals(key0, key1);
+                    }
+                }
+#endif
+            }
+            else
+            {
+                return customComparer.Equals(key0, key1);
+            }
         }
 
-        void ICollection<KeyValuePair<TKey, TValue>>.CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int KeyHashCode(TKey key)
         {
-            CopyTo(array, index);
+            int result;
+            IEqualityComparer<TKey> customComparer = _customComparer;
+            if (customComparer == null)
+            {
+                if (default(TKey) == null)
```

Comment:
This is not necessary. `result = key?.GetHashCode() ?? 0;` should work well for all cases.

Reply:
Also key is checked everywhere before being added; and throws ArgumentNullExceptions so can just be `key.GetHashCode()`?